File: controllers/attendanceController.js
// =====================================================
// Controller: Chấm công
// =====================================================
const AttendanceModel = require('../models/attendanceModel');

const getAll = async (req, res) => {
    try {
        const options = {
            page: parseInt(req.query.page) || 1,
            limit: parseInt(req.query.limit) || 20,
            maNhanVien: req.query.maNhanVien || '',
            thang: req.query.thang || '',
            nam: req.query.nam || '',
            search: req.query.search || ''
        };

        const [rows, total] = await Promise.all([
            AttendanceModel.getAll(options),
            AttendanceModel.count(options)
        ]);

        return res.json({
            success: true,
            data: rows,
            pagination: {
                page: options.page,
                limit: options.limit,
                total,
                totalPages: Math.ceil(total / options.limit)
            }
        });
    } catch (error) {
        return res.status(500).json({ success: false, message: 'Lỗi lấy dữ liệu chấm công' });
    }
};

const getOne = async (req, res) => {
    try {
        const record = await AttendanceModel.findById(req.params.id);
        if (!record) return res.status(404).json({ success: false, message: 'Không tìm thấy bản ghi' });
        return res.json({ success: true, data: record });
    } catch (error) {
        return res.status(500).json({ success: false, message: 'Lỗi máy chủ' });
    }
};

const create = async (req, res) => {
    try {
        const data = req.body;
        // Tự tính số giờ làm
        if (data.gioVao && data.gioRa && !data.soGioLam) {
            const [h1, m1] = data.gioVao.split(':').map(Number);
            const [h2, m2] = data.gioRa.split(':').map(Number);
            data.soGioLam = ((h2 * 60 + m2) - (h1 * 60 + m1)) / 60;
        }
        const id = await AttendanceModel.create(data);
        const record = await AttendanceModel.findById(id);
        return res.status(201).json({ success: true, message: 'Thêm chấm công thành công', data: record });
    } catch (error) {
        return res.status(500).json({ success: false, message: 'Lỗi thêm chấm công' });
    }
};

const update = async (req, res) => {
    try {
        const data = req.body;
        if (data.gioVao && data.gioRa && !data.soGioLam) {
            const [h1, m1] = data.gioVao.split(':').map(Number);
            const [h2, m2] = data.gioRa.split(':').map(Number);
            data.soGioLam = ((h2 * 60 + m2) - (h1 * 60 + m1)) / 60;
        }
        await AttendanceModel.update(req.params.id, data);
        const record = await AttendanceModel.findById(req.params.id);
        return res.json({ success: true, message: 'Cập nhật thành công', data: record });
    } catch (error) {
        return res.status(500).json({ success: false, message: 'Lỗi cập nhật chấm công' });
    }
};

const remove = async (req, res) => {
    try {
        await AttendanceModel.delete(req.params.id);
        return res.json({ success: true, message: 'Xóa thành công' });
    } catch (error) {
        return res.status(500).json({ success: false, message: 'Lỗi xóa' });
    }
};

module.exports = { getAll, getOne, create, update, remove };

File: quan.comdemo/quan.comdemo/src/main/java/K23/CNT1/_5/DAQ/quan/comdemo/controller/AttendanceController.java
package K23.CNT1._5.DAQ.quan.comdemo.controller;

import K23.CNT1._5.DAQ.quan.comdemo.dto.ApiResponse;
import K23.CNT1._5.DAQ.quan.comdemo.entity.ChamCong;
import K23.CNT1._5.DAQ.quan.comdemo.service.AttendanceService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/attendance")
public class AttendanceController {

    @Autowired
    private AttendanceService attendanceService;

    @GetMapping
    public ResponseEntity<ApiResponse<List<ChamCong>>> getAll(
            @RequestParam(defaultValue = "1") int page,
            @RequestParam(defaultValue = "20") int limit,
            @RequestParam(required = false) Integer maNhanVien,
            @RequestParam(required = false) Integer thang,
            @RequestParam(required = false) Integer nam,
            @RequestParam(required = false, defaultValue = "") String search) {

        Page<ChamCong> resultPage = attendanceService.getAll(page, limit, maNhanVien, thang, nam, search);

        Map<String, Object> pagination = new HashMap<>();
        pagination.put("page", page);
        pagination.put("limit", limit);
        pagination.put("total", resultPage.getTotalElements());
        pagination.put("totalPages", resultPage.getTotalPages());

        return ResponseEntity.ok(ApiResponse.ok(resultPage.getContent(), pagination));
    }

    @GetMapping("/{id}")
    public ResponseEntity<ApiResponse<ChamCong>> getOne(@PathVariable Integer id) {
        ChamCong cc = attendanceService.getOne(id);
        if (cc == null) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND).body(ApiResponse.error("Không tìm thấy bản ghi"));
        }
        return ResponseEntity.ok(ApiResponse.ok(cc));
    }

    @PostMapping
    public ResponseEntity<ApiResponse<ChamCong>> create(@RequestBody Map<String, Object> body) {
        try {
            ChamCong created = attendanceService.create(body);
            return ResponseEntity.status(HttpStatus.CREATED).body(ApiResponse.ok("Thêm chấm công thành công", created));
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(ApiResponse.error("Lỗi thêm chấm công: " + e.getMessage()));
        }
    }

    @PutMapping("/{id}")
    public ResponseEntity<ApiResponse<ChamCong>> update(@PathVariable Integer id, @RequestBody Map<String, Object> body) {
        try {
            ChamCong updated = attendanceService.update(id, body);
            return ResponseEntity.ok(ApiResponse.ok("Cập nhật thành công", updated));
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(ApiResponse.error("Lỗi cập nhật chấm công: " + e.getMessage()));
        }
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<ApiResponse<String>> delete(@PathVariable Integer id) {
        try {
            attendanceService.delete(id);
            return ResponseEntity.ok(ApiResponse.ok("Xóa thành công", null));
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(ApiResponse.error("Lỗi xóa: " + e.getMessage()));
        }
    }
}
